dev_setup.sh

```diff
@@ -0,0 +1,189 @@
+
+# Create a new, empty file for environment variables
+VAR_FILE=$ROOT/env.sh
+truncate -s 0 $VAR_FILE
+echo "# Generated environment variables. Manual changes will get clobbered by dev_setup.sh" >> $VAR_FILE
+
+WHAT4_SOLVERS_SNAPSHOT="snapshot-20240212"
+WHAT4_SOLVERS_URL="https://github.com/GaloisInc/what4-solvers/releases/download/$WHAT4_SOLVERS_SNAPSHOT/"
+WHAT4_SOLVERS_MACOS_12="macos-12-X64-bin.zip"
+WHAT4_SOLVERS_MACOS_14="macos-14-ARM64-bin.zip"
+WHAT4_CVC4_VERSION="version 1.8"
+WHAT4_CVC5_VERSION="version 1.1.1"
+
+# Set of supported platforms:
+MACOS14="macos14"
+MACOS12="macos12" # actually, this isn't supported yet
+
+# Returns a string indicating the platform (from the set above), or empty if
+# a supported platform is not detected.
+function supported_platform {
+    # Make sure we're running on a supported platform
+    case $(uname -s) in
+        Darwin)
+            if [ $(uname -m) = "arm64" ]; then
+                echo $MACOS14
+            # This is how we'd support macOS 12. Since this hasn't been tested yet,
+            # we withhold official support.
+            # This might bork on something running macOS <12, since we're basing
+            # the it on the hardware, not the specific version.
+            elif [ $(uname -m) = "x86_64" ]; then
+                # CRYPTOL_PLATFORM=$MACOS12
+                echo ""
+            fi;;
+        *) echo ""
+    esac
+}
+
+RED="\033[0;31m"
+function notice {
+    echo "[NOTICE] $*"
+}
+
+function is_installed {
+    command -v $* >/dev/null 2>&1
+}
+
+# Requires: LOG set to log file path.
+function logged {
+    if ! [ -z "$LOG" ]
+    then
+        # This may or may not be the same directory that this script lives in.
+        mkdir -p `dirname $LOG`
+        echo "$@" >>$LOG
+        if ! "$@" >>$LOG 2>&1
+        then
+            echo
+            tail -n 50 $LOG
+            echo
+            echo "[ERROR] An error occurred; please see $LOG"
+            echo "[ERROR] The last 50 lines are printed above for convenience"
+            exit 1
+        fi
+    fi
+}
+
+function update_submodules {
+    cd $ROOT
+    notice "Updating submodules"
+    logged git submodule update --init
+}
+
+function install_ghcup {
+    if ! is_installed ghcup
+    then
+        ghcup_url="https://get-ghcup.haskell.org"
+        notice "ghcup not found; do you want to install from $ghcup_url"
+        read -p "Press Enter to continue or 'n' to skip: " consent
+        case $consent in
+            [Nn]* ) notice "Skipping ghcup installation"; return;;
+            * ) notice "Installing ghcup";;
+        esac
+
+        # Technically the installation only requires cabal, but it's
+        # recommended to get the whole GCH shebang in one package.
+        # The output is not routed to log because the installer is interactive.
+        ghc_dir=$(mktemp -d)
+        logged curl --proto '=https' --tlsv1.2 -sSf -o "$ghc_dir/ghcup.bin" $ghcup_url
+        cd $ghc_dir
+        sh ghcup.bin
```

Comment:
I don't think anything ever cds back. Might want to run it in a subshell so that the change is discarded when done. (That is, ```(cd $ghc_dir && sh ghcup.bin)```)

Reply:
This is now gone but I had done the same thing in the what4 solvers bit. That's now fixed.